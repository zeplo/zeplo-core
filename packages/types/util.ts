
export type Maybe<T> = T | null

export type JsonValue =  boolean | number | string | null | JsonArray | JsonMap
export interface JsonMap { [key: string]: JsonValue }
export interface JsonArray extends Array<JsonValue> {}

export type DeepPartial<T> = T extends Function ? T : (T extends object ? { [P in keyof T]?: DeepPartial<T[P]>; } : T);

/**
 * Get all of the keys to which U can be assigned.
 */
export type OnlyKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? K : never
}[keyof T];

/**
 * Get the interface containing only properties to which U can be assigned.
 */
export type OnlyUndefined<T> = {
  [K in OnlyKeys<T, undefined>]: T[K]
}

/**
 * Get all of the keys except those to which U can be assigned.
 */
export type ExcludeKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? never : K
}[keyof T];

/**
 * Get the interface containing no properties to which U can be assigned.
 */
export type ExcludeUndefined<T> = {
  [K in ExcludeKeys<T, undefined>]: T[K]
}

/**
 * Get the interface where all properties are optional.
 */
export type Optional<T> = {[K in keyof T]?: T[K]};

/**
 * Get the interface where properties that can be assigned undefined are
 * also optional.
 */
export type UndefinedOptional<T> = ExcludeUndefined<T> & Optional<OnlyUndefined<T>>

export type DeepUndefinedOptional<T> = T extends (...args: any[]) => any
  ? T
  : T extends any[]
  ? _DeepRequiredArray<T[number]>
  : T extends object
  ? _DeepRequiredObject<UndefinedOptional<T>>
  : T;

 export interface _DeepRequiredArray<T>
  extends Array<DeepUndefinedOptional<T>> {}

  export type _DeepRequiredObject<T> = {
  [P in keyof T]: DeepUndefinedOptional<T[P]>
}