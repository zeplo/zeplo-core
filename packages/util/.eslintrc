{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "parser":  "@typescript-eslint/parser",
  "env": {
    "browser": true,
    "node": true,
    "es6": true,
    "mocha": true,
    "jest": true
  },
  "rules": {
    "semi": ["error", "never"],
    "no-underscore-dangle": 0,
    "space-before-function-paren": ["error", "always"],
    "arrow-body-style": 0,
    "no-use-before-define": ["error", { "functions": false }],
    "arrow-parens": 0,
  }
}
