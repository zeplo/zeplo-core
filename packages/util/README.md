# @zeplo/util

Utility functions that are used across packages.