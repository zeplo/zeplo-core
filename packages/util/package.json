{
  "name": "@zeplo/util",
  "version": "2.1.0",
  "main": "dist/index.js",
  "files": [
    "dist/*"
  ],
  "license": "MIT",
  "scripts": {
    "clean": "rimraf dist",
    "dev": "env-cmd ts-node",
    "build": "tsc",
    "prepublish": "yarn clean && yarn build"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/fs-extra": "^9.0.13",
    "@types/lodash": "^4.14.178",
    "@types/node": "^17.0.18",
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "env-cmd": "^10.1.0",
    "eslint": "^8.9.0",
    "rimraf": "^3.0.2",
    "ts-node": "^10.5.0",
    "ts-node-dev": "^1.1.8",
    "typescript": "^4.5.5"
  },
  "dependencies": {
    "@zeplo/errors": "^2.0.8",
    "@zeplo/types": "^2.1.0",
    "express": "^4.17.3",
    "express-async-handler": "^1.2.0",
    "lodash": "^4.17.21",
    "uuid": "^8.3.2"
  }
}
