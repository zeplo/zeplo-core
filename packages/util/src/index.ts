export { cleanQuery, cleanHeaders } from './clean'
export { 
  parseRequest, 
  parseBody, 
  getBackoff, 
  getStartTimeFromFeatures,
  getRawFeatures,
  parseRawRequestObject, 
  createRequestJob ,
  getValueFromQueryOrHeader,
} from './parse'
export { 
  createStepsJobs,
  hasCyclicalDependency,
} from './steps'
export { 
  requestHandler,
  jsonRequestHandler,
  queryToString,
  queryToNumber,
  parseQuery,
} from './request'