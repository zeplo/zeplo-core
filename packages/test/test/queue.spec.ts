/**
 * @jest-environment node
 */

import { omit } from 'lodash'
import qs from 'query-string'
import FormData from 'form-data'
import fs from 'fs'
import path from 'path'
import { createWorkspace } from '#/util/data'
import waitFor from '#/util/waitFor'
import queue from '#/util/queue'

jest.setTimeout(60000)

describe('queue.spec', () => {
  const workspace = 'test-workspace-queue'
  const token = workspace

  let cleanup: () => Promise<void>
  beforeAll(async () => {
    cleanup = await createWorkspace(workspace, token)
  })
  afterAll(() => cleanup())

  test('get server info', async () => {
    const res = await queue.get('/')
    expect(res.status).toBe(200)
    // expect(res.data.name).toBe('@zeplo/router')
    // expect(res.data.home).toBe('https://zeplo.io')
  })

  test('queue GET request', async () => {
    const res = await queue.get(`/https://postman-echo.com/get?_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'GET' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const respBody = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(respBody.data).toMatchObject({
      url: 'https://postman-echo.com/get',
    })
  })

  test('queue GET request without scheme', async () => {
    const res = await queue.get(`/postman-echo.com/get?_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'GET' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('queue POST request', async () => {
    const res = await queue.post(`/https://postman-echo.com/post?_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'POST' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('queue DELETE request', async () => {
    const res = await queue.delete(`/https://postman-echo.com/delete?_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'DELETE' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('queue PUT request', async () => {
    const res = await queue.put(`/https://postman-echo.com/put?_token=${token}`)
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'PUT' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('queue PATCH request', async () => {
    const res = await queue.patch(`/https://postman-echo.com/patch?_token=${token}`, {
      x: 1,
    })
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'PATCH' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('queue POST request with url-enconded', async () => {
    const res = await queue.post(`/https://postman-echo.com/post?_token=${token}`,
      qs.stringify({ abc: '1' }),
      { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } })
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'POST' },
      response: { status: 200 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      form: { abc: '1' },
      url: 'https://postman-echo.com/post',
    })
  })

  test('queue POST request with multi-part form', async () => {
    const file = fs.readFileSync(path.resolve(__dirname, './avatar.png'))
    const form = new FormData()
    form.append('abc', '1')
    form.append('icon', file)

    const res = await queue.post(`/https://postman-echo.com/post?_token=${token}`,
      form.getBuffer(),
      { headers: form.getHeaders() })
    expect(res.data).toHaveProperty('id')
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'SUCCESS',
      source: 'REQUEST',
      request: { method: 'POST' },
      response: { status: 200 },
    })

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      files: { undefined: `data:application/octet-stream;base64,${file.toString('base64')}` },
      form: { abc: '1' },
      url: 'https://postman-echo.com/post',
    })
  })

  test('queue POST request with header, params and JSON body', async () => {
    const res = await queue.post(`/https://postman-echo.com/post?x=1&_token=${token}`, { abc: 1 }, {
      params: { y: 2 },
      headers: { 'x-custom-header': '123' },
    })
    const id = res.data.id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        headers: { 'x-custom-header': '123', 'content-type': 'application/json' },
        params: { x: '1', y: '2' },
        hasbody: true,
      },
    })
    expect(qs.parseUrl(req.data.request.url)).toMatchObject({
      url: 'https://postman-echo.com/post',
      query: { x: '1', y: '2' },
    })
    delete req.data.request.url
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const reqBody = await queue.get(`/requests/${id}/request.body?_token=${token}`)
    expect(reqBody.data).toMatchObject({ abc: 1 })

    const respBody = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(respBody.data).toMatchObject({
      headers: { 'x-custom-header': '123' },
      json: { abc: 1 },
      args: { y: '2', x: '1' },
    })
    expect(qs.parseUrl(respBody.data.url)).toMatchObject({
      url: 'https://postman-echo.com/post',
      query: { x: '1', y: '2' },
    })
  })

  test('queue GET request with error', async () => {
    const res = await queue.get(`/https://postman-echo.com/post?_token=${token}`)
    expect(res.data).toHaveProperty('id')

    const req = await waitFor(
      () => queue.get(`/requests/${res.data.id}?_token=${token}`),
      ({ data }) => data.status === 'ERROR',
    )
    expect(req.data).toMatchObject({
      attempts: 1,
      status: 'ERROR',
      source: 'REQUEST',
      request: { method: 'GET' },
      response: { status: 404 },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  xtest('errors if no token provided', async () => {
    const error = await queue.get('/https://postman-echo.com').catch((err) => err)
    expect(error.response.status).toBe(401)
  })
})
