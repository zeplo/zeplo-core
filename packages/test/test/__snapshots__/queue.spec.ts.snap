// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queue.spec queue DELETE request 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "DELETE",
  "params": Object {},
  "path": "/delete",
  "scheme": "https",
  "url": "https://postman-echo.com/delete",
}
`;

exports[`queue.spec queue GET request 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;

exports[`queue.spec queue GET request with error 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`queue.spec queue GET request without scheme 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;

exports[`queue.spec queue PATCH request 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/json",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "PATCH",
  "params": Object {},
  "path": "/patch",
  "scheme": "https",
  "url": "https://postman-echo.com/patch",
}
`;

exports[`queue.spec queue POST request 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/x-www-form-urlencoded",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`queue.spec queue POST request with header, params and JSON body 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/json",
    "user-agent": "axios/0.26.0",
    "x-custom-header": "123",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {
    "x": "1",
    "y": "2",
  },
  "path": "/post",
  "scheme": "https",
}
`;

exports[`queue.spec queue POST request with url-enconded 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/x-www-form-urlencoded",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`queue.spec queue PUT request 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "content-type": "application/x-www-form-urlencoded",
    "user-agent": "axios/0.26.0",
  },
  "host": "postman-echo.com",
  "method": "PUT",
  "params": Object {},
  "path": "/put",
  "scheme": "https",
  "url": "https://postman-echo.com/put",
}
`;
